server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java

```diff
@@ -48,63 +49,140 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 /**
  * A {@link RoutingNode} represents a cluster node associated with a single {@link DiscoveryNode} including all shards
  * that are hosted on that nodes. Each {@link RoutingNode} has a unique node id that can be used to identify the node.
  */
 public class RoutingNode implements Iterable<ShardRouting> {
 
+    static class BucketedShards implements Iterable<ShardRouting> {
+        private final Tuple<LinkedHashMap<ShardId, ShardRouting>, LinkedHashMap<ShardId, ShardRouting>> shardTuple; // LinkedHashMap to
+                                                                                                                    // preserve order
+
+        BucketedShards(LinkedHashMap<ShardId, ShardRouting> primaryShards, LinkedHashMap<ShardId, ShardRouting> replicaShards) {
+            this.shardTuple = new Tuple(primaryShards, replicaShards);
+        }
+
+        public boolean isEmpty() {
+            return this.shardTuple.v1().isEmpty() && this.shardTuple.v2().isEmpty();
+        }
+
+        public int size() {
+            return this.shardTuple.v1().size() + this.shardTuple.v2().size();
+        }
+
+        public boolean containsKey(ShardId shardId) {
+            return this.shardTuple.v1().containsKey(shardId) || this.shardTuple.v2().containsKey(shardId);
+        }
+
+        public ShardRouting get(ShardId shardId) {
+            if (this.shardTuple.v1().containsKey(shardId)) {
+                return this.shardTuple.v1().get(shardId);
+            }
+            return this.shardTuple.v2().get(shardId);
+        }
+
+        public ShardRouting add(ShardRouting shardRouting) {
+            return put(shardRouting.shardId(), shardRouting);
+        }
+
+        public ShardRouting put(ShardId shardId, ShardRouting shardRouting) {
+            ShardRouting ret;
+            if (shardRouting.primary()) {
+                ret = this.shardTuple.v1().put(shardId, shardRouting);
+                if (this.shardTuple.v2().containsKey(shardId)) {
+                    ret = this.shardTuple.v2().remove(shardId);
+                }
+            } else {
+                ret = this.shardTuple.v2().put(shardId, shardRouting);
+                if (this.shardTuple.v1().containsKey(shardId)) {
+                    ret = this.shardTuple.v1().remove(shardId);
+                }
+            }
+
+            return ret;
+        }
+
+        public ShardRouting remove(ShardId shardId) {
+            if (this.shardTuple.v1().containsKey(shardId)) {
+                return this.shardTuple.v1().remove(shardId);
+            }
+            return this.shardTuple.v2().remove(shardId);
+        }
+
+        @Override
+        public Iterator<ShardRouting> iterator() {
+            final Iterator<ShardRouting> primaryIterator = Collections.unmodifiableCollection(this.shardTuple.v1().values()).iterator();
```

Comment:
Nitpick, but I think the Java Stream API can do this for you. Something like:

```
Stream.concat(
    Collections.unmodifiableCollection(this.shardTuple.v1().values()).stream(),
    Collections.unmodifiableCollection(this.shardTuple.v2().values()).stream()
).iterator();
```

Reply:
Sure, will include as part of documentation change